pkg/predicate/predicate.go

```diff
@@ -20,33 +20,31 @@ import (
 	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	appsv1 "k8s.io/api/apps/v1"
 	coordinationv1 "k8s.io/api/coordination/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
 	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
 )
 
+func hasOperationAnnotation(obj client.Object) bool {
+	return obj.GetAnnotations()[v1beta1constants.GardenerOperation] == v1beta1constants.GardenerOperationReconcile
+}
+
 // HasOperationAnnotation is a predicate for the operation annotation.
 func HasOperationAnnotation() predicate.Predicate {
-	f := func(obj runtime.Object) bool {
-		etcd, ok := obj.(*druidv1alpha1.Etcd)
-		if !ok {
-			return false
-		}
```

Comment:
Why aren't we retaining the nil check on the etcd object here? If the object is `nil`, wouldn't `hasOperationAnnotation` call fail?

Reply:
Actually, this is not a nil check. If you pass the following, the code will panic also with the previous code in place:
```
var etcd *druidv1alpha1.Etcd = nil

createEvent := event.CreateEvent{Object: etcd}
```
So, I wondered about the benefit of having this check at all?

Reply:
Oh ok. I was just intrigued since the predicates returned in `EtcdReconciliationFinished` had these checks in place. I still feel it's better to keep this check, to be able to return `false` early rather than panic. WDYT?

Reply:
> I was just intrigued since the predicates returned in EtcdReconciliationFinished had these checks in place.

There it's needed because we access a particular field of the `etcd` resource which is `etcd.status.ready`.

Here, we only need to get the annotations which is possible via the `client.Object` interface that is guaranteed through the `{Create,Update,Delete,Generic}Event` object.

A panic is not prevented by the earlier check because you actually need a [typed nil check](https://stackoverflow.com/questions/13476349/check-for-nil-and-nil-interface-in-go) for this which has never been used.